src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObjectNative.cs

```diff
@@ -0,0 +1,380 @@
+                }
+            }
+            return _sniPacket;
+        }
+
+        internal override void ClearAllWritePackets()
+        {
+            if (_sniPacket != null)
+            {
+                _sniPacket.Dispose();
+                _sniPacket = null;
+            }
+            lock (_writePacketLockObject)
+            {
+                Debug.Assert(_pendingWritePackets.Count == 0 && _asyncWriteCount == 0, "Should not clear all write packets if there are packets pending");
+                _writePacketCache.Clear();
+            }
+        }
+
+        internal override void SetPacketData(object packet, byte[] buffer, int bytesUsed) 
+            => SNINativeMethodWrapper.SNIPacketSetData((SNIPacket)packet, buffer, bytesUsed);
+
+        internal override uint SniGetConnectionId(ref Guid clientConnectionId) 
+            => SNINativeMethodWrapper.SniGetConnectionId(Handle, ref clientConnectionId);
+
+        internal override uint DisabeSsl() 
+            => SNINativeMethodWrapper.SNIRemoveProvider(Handle, SNINativeMethodWrapper.ProviderEnum.SSL_PROV);
+
+        internal override uint EnableMars(ref uint info) 
+            => SNINativeMethodWrapper.SNIAddProvider(Handle, SNINativeMethodWrapper.ProviderEnum.SMUX_PROV, ref info);
+
+        internal override uint EnableSsl(ref uint info)
+        {
+            // Add SSL (Encryption) SNI provider.
+            return SNINativeMethodWrapper.SNIAddProvider(Handle, SNINativeMethodWrapper.ProviderEnum.SSL_PROV, ref info);
+        }
+
+        internal override uint SetConnectionBufferSize(ref uint unsignedPacketSize) 
+            => SNINativeMethodWrapper.SNISetInfo(Handle, SNINativeMethodWrapper.QTypes.SNI_QUERY_CONN_BUFSIZE, ref unsignedPacketSize);
+
+        internal override uint GenerateSspiClientContext(byte[] receivedBuff, uint receivedLength, byte[] sendBuff, ref uint sendLength, byte[] _sniSpnBuffer) 
+            => SNINativeMethodWrapper.SNISecGenClientContext(Handle, receivedBuff, receivedLength, sendBuff, ref sendLength, _sniSpnBuffer);
+        
+
+        internal override uint WaitForSSLHandShakeToComplete() 
+            => SNINativeMethodWrapper.SNIWaitForSSLHandshakeToComplete(Handle, GetTimeoutRemaining());
+
+        internal override void DisposePacketCache()
+        {
+            if (_writePacketCache != null)
+            {
+                lock (_writePacketLockObject)
+                {
+                    try { }
+                    finally
+                    {
+                        _writePacketCache.Dispose();
+                        // Do not set _writePacketCache to null, just in case a WriteAsyncCallback completes after this point
+                    }
+                }
+            }
+        }
+
+        internal override void DisposeHandle()
+        {
+            var sessionHandle = Handle;
+            if (sessionHandle != null)
+            {
+                sessionHandle.Dispose();
+            }
```

Comment:
I think I can get rid of DisposeHandle() This is called by BestEffortCleanUp from the TdsParserSessionPool.cs which is not called by any caller. 
The BestEffortCleanUp existed in NetFx for Thread Aborts.